Cooperative request intercepts (#6735)
puppeteer/puppeteer: docs/api.md

  * [elementHandle.type(text[, options])](#elementhandletypetext-options)
  * [elementHandle.uploadFile(...filePaths)](#elementhandleuploadfilefilepaths)
- [class: HTTPRequest](#class-httprequest)
  * [httpRequest.abort([errorCode])](#httprequestaborterrorcode)
  * [httpRequest.abort([errorCode], [priority])](#httprequestaborterrorcode-priority)
  * [httpRequest.abortErrorReason()](#httprequestaborterrorreason)
  * [httpRequest.continue([overrides])](#httprequestcontinueoverrides)
  * [httpRequest.continue([overrides], [priority])](#httprequestcontinueoverrides-priority)
  * [httpRequest.continueRequestOverrides()](#httprequestcontinuerequestoverrides)
  * [httpRequest.enqueueInterceptAction(handler)](#httprequestnequeueinterceptactionhandler)
  * [httpRequest.enqueueInterceptAction(pendingHandler)](#httprequestenqueueinterceptactionpendinghandler)
  * [httpRequest.failure()](#httprequestfailure)
  * [httpRequest.finalizeInterceptions()](#httprequestfinalizeinterceptions)
  * [httpRequest.frame()](#httprequestframe)
  * [httpRequest.headers()](#httprequestheaders)

>   * [httpRequest.interceptResolution()](#httprequestinterceptresolution)

Author: Did you mean to remove this? It’s an important method for package authors.

- returns: <[Accessibility]>

> #### page.abortErrorReason()

Reviewer: Why do we need this method on `page`?

Author: Doc error :D

    DEFAULT_INTERCEPT_RESOLUTION_PRIORITY + 5
  ); // Cooperative behavior: responds at priority 5
});
page.on('request', (request) => {
  console.log(request.interceptResolution()); // ['respond', 15], because respond @ 15 > continue @ 15 > abort @ 10
});
```

##### Upgrading to Cooperative Mode for Package Maintainers

If you are package maintainer and your package uses intercept handlers, you can update your intercept handlers to use Cooperative Mode. Suppose you have the following existing handler:

```ts
page.on('request', (interceptedRequest) => {
  if (
    interceptedRequest.url().endsWith('.png') ||
    interceptedRequest.url().endsWith('.jpg')
  )
    interceptedRequest.abort();
  else interceptedRequest.continue();
});
```

To use Cooperative Mode, upgrade `continue()` and `abort()`:

```ts
page.on('request', (interceptedRequest) => {
  if (
    interceptedRequest.url().endsWith('.png') ||
    interceptedRequest.url().endsWith('.jpg')
  )
    interceptedRequest.abort('failed', DEFAULT_INTERCEPT_RESOLUTION_PRIORITY);
  else
    interceptedRequest.continue(
      interceptedRequest.continueRequestOverrides(),
      DEFAULT_INTERCEPT_RESOLUTION_PRIORITY
    );
});
```

> With those simiple upgrades, your handler will use Cooperative Mode instead.

Reviewer: simple

})();
```

##### Cooperative Intercept Mode and Legacy Intercept Mode

`request.respond`, `request.abort`, and `request.continue` can accept an optional `priority` to activate Cooperative Intercept Mode. In Cooperative Mode, all intercept handlers are guaraneed to run and the interception will be resolved to the highest priority resolution. Here are the rules of Cooperative Mode:

- Async handlers will finish before intercept resolution is finalized
- The highest priority interception resolution "wins", ie, the interception will ultimately be aborted/responded/continued according to which resolution was given the highest priority.
- In the event of a tie, `abort` > `respond` > `continue`

> For standardization, when specifying a Cooperative Mode priority use `DEFAULT_INTERCEPT_RESOLUTION_PRIORITY` (currently 10) unless you have a clear reason to use a higher priority. This will gracefully prefer `respond` over `continue` and `abort` over `respond`. If you do intentionally want to use a higher priority, `DEFAULT_INTERCEPT_RESOLUTION_PRIORITY + n` is the recommended approach. For example, `continue({}, DEFAULT_INTERCEPT_RESOLUTION_PRIORITY + 4)` rather than `continue({}, 14)`.

Reviewer: I'm wondering if there is a way to get around carrying the `DEFAULT_INTERCEPT_RESOLUTION_PRIORITY` constant around all the time. I feel like this makes for a rather bumpy experience. WDYT about changing the default value to 0 and allowing both negative and positive priorities or instead fixing it to some given value like 10?

})();
```

##### Cooperative Intercept Mode and Legacy Intercept Mode

> `request.respond`, `request.abort`, and `request.continue` can accept an optional `priority` to activate Cooperative Intercept Mode. In Cooperative Mode, all intercept handlers are guaraneed to run and the interception will be resolved to the highest priority resolution. Here are the rules of Cooperative Mode:

Reviewer 2: ```suggestion
`request.respond`, `request.abort`, and `request.continue` can accept an optional `priority` to activate Cooperative Intercept Mode. In Cooperative Mode, all intercept handlers are guaranteed to run and the interception is resolved to the highest-priority resolution. Here are the rules of Cooperative Mode:
```

> - Async handlers will finish before intercept resolution is finalized
> - The highest priority interception resolution "wins", ie, the interception will ultimately be aborted/responded/continued according to which resolution was given the highest priority.
> - In the event of a tie, `abort` > `respond` > `continue`
> 
> For standardization, when specifying a Cooperative Mode priority use `0` unless you have a clear reason to use a higher priority. This will gracefully prefer `respond` over `continue` and `abort` over `respond`. If you do intentionally want to use a different priority, higher priorities win over lower priorities. Negative priorities are allowed. For example, `continue({}, 4)` would win over `continue({}, -2)`.
> 
> To preserve backward compatibility, any handler resolving the intercept without specifying `priority` (Legacy Mode) will cause immediate resolution. For Cooperative Mode to work, all resolutions must use a `priority`.

Reviewer 2: ```suggestion
- Async handlers finish before intercept resolution is finalized.
- The highest priority interception resolution "wins", i.e. the interception is ultimately aborted/responded/continued according to which resolution was given the highest priority.
- In the event of a tie, `abort` > `respond` > `continue`.

For standardization, when specifying a Cooperative Mode priority use `0` unless you have a clear reason to use a higher priority. This gracefully prefers `respond` over `continue` and `abort` over `respond`. If you do intentionally want to use a different priority, higher priorities win over lower priorities. Negative priorities are allowed. For example, `continue({}, 4)` would win over `continue({}, -2)`.

To preserve backward compatibility, any handler resolving the intercept without specifying `priority` (Legacy Mode) causes immediate resolution. For Cooperative Mode to work, all resolutions must use a `priority`.
```

  request.abort('failed', 0); // Cooperative behavior: aborts at priority 10
});
page.on('request', (request) => {
  request.continue(request.continueRequestOverrides(), 5); // Cooperative behavior: continues at priority 5
});
page.on('request', (request) => {
  console.log(request.interceptResolution()); // ['continue', 5], because continue @ 5 > abort @ 0
});
```

In this example, Cooperative Mode is active because all handlers specify `priority`. `respond()` wins because its priority ties with `continue()`, but `respond()` beats `continue()`.

```ts
// Final outcome: cooperative respond()
page.setRequestInterception(true);
page.on('request', (request) => {
  request.abort('failed', 10); // Cooperative behavior: aborts at priority 10
});
page.on('request', (request) => {
  request.continue(request.continueRequestOverrides(), 15); // Cooperative behavior: continues at priority 15
});
page.on('request', (request) => {
  request.respond(request.responseForRequest(), 15); // Cooperative behavior: responds at priority 15
});
page.on('request', (request) => {
  console.log(request.interceptResolution()); // ['respond', 15], because respond @ 15 > continue @ 15 > abort @ 10
});
```

> ##### Upgrading to Cooperative Mode for Package Maintainers

Reviewer 2: We use sentence case for titles:

```suggestion
##### Upgrading to Cooperative Mode for package maintainers
```

("Cooperative Mode" is consistently capitalized even in sentences)

If you are package maintainer and your package uses intercept handlers, you can update your intercept handlers to use Cooperative Mode. Suppose you have the following existing handler:

```ts
page.on('request', (interceptedRequest) => {
  if (
    interceptedRequest.url().endsWith('.png') ||
    interceptedRequest.url().endsWith('.jpg')
  )
    interceptedRequest.abort();
  else interceptedRequest.continue();
});
```

To use Cooperative Mode, upgrade `continue()` and `abort()`:

```ts
page.on('request', (interceptedRequest) => {
  if (
    interceptedRequest.url().endsWith('.png') ||
    interceptedRequest.url().endsWith('.jpg')
  )
    interceptedRequest.abort('failed', 0);
  else
    interceptedRequest.continue(
      interceptedRequest.continueRequestOverrides(),
      0
    );
});
```

> With those simple upgrades, your handler will use Cooperative Mode instead.

Reviewer 2: ```suggestion
With those simple upgrades, your handler now uses Cooperative Mode instead.
```

  if (
    interceptedRequest.url().endsWith('.png') ||
    interceptedRequest.url().endsWith('.jpg')
  )
    interceptedRequest.abort('failed', _priority);
  else
    interceptedRequest.continue(
      interceptedRequest.continueRequestOverrides(),
      _priority
    );
});
```

If your package calls for more fine-grained control resolution priorities, use a config pattern like this:

```ts
interface ResolutionStrategy {
  abortPriority: number;
  continuePriority: number;
}
const DEFAULT_STRATEGY: ResolutionStrategy = {
  abortPriority: 0,
  continuePriority: 0,
};

// Defaults to undefined which preserves legacy mode behavior
let _strategy: Partial<ResolutionStrategy> = {};

export const setInterceptResolutionStrategy = (strategy: ResolutionStrategy) =>
  (_strategy = { ...DEFAULT_STRATEGY, ...strategy });

page.on('request', (interceptedRequest) => {
  if (
    interceptedRequest.url().endsWith('.png') ||
    interceptedRequest.url().endsWith('.jpg')
  )
    interceptedRequest.abort('failed', _strategy.abortPriority);
  else
    interceptedRequest.continue(
      interceptedRequest.continueRequestOverrides(),
      _strategy.continuePriority
    );
});
```

> The above solution will ensure that your package continues to work as expected with other legacy handlers while also allowing the user to adjust the importance of your package in the resolution chain when Cooperative Mode is being used.

Reviewer 2: ```suggestion
The above solution ensures that your package continues to work as expected with other legacy handlers while also allowing the user to adjust the importance of your package in the resolution chain when Cooperative Mode is being used.
```

  - `namenotresolved` - The host name could not be resolved.
  - `timedout` - An operation timed out.
  - `failed` - A generic failure occurred.

> - `priority` <[number]> - Optional intercept abort priority. If provided, intercept will be resolved using coopeative handling rules. Otherwise, intercept will be resovled immediately.

Reviewer 2: ```suggestion
- `priority` <[number]> - Optional intercept abort priority. If provided, intercept is resolved using cooperative handling rules. Otherwise, intercept is resolved immediately.
```

- returns: <[Promise]>

Aborts request. To use this, request interception should be enabled with `page.setRequestInterception`.
Exception is immediately thrown if the request interception is not enabled.

#### httpRequest.abortErrorReason()

- returns: <[string]> of type [Protocol.Network.ErrorReason](https://chromedevtools.github.io/devtools-protocol/tot/Network/#type-ErrorReason).

Returns the most recent reason for aborting set by the previous call to abort() in Cooperative Mode.

#### httpRequest.continue([overrides], [priority])

- `overrides` <[Object]> Optional request overwrites, which can be one of the following:
  - `url` <[string]> If set changes the request URL. This is not a redirect. The request will be silently forwarded to the new URL. For example, the address bar will show the original URL.
  - `method` <[string]> If set changes the request method (e.g. `GET` or `POST`).
  - `postData` <[string]> If set changes the post data of request.
  - `headers` <[Object]> If set changes the request HTTP headers. Header values will be converted to a string.

> - `priority` <[number]> - Optional intercept abort priority. If provided, intercept will be resolved using coopeative handling rules. Otherwise, intercept will be resovled immediately.

Reviewer 2: ```suggestion
- `priority` <[number]> - Optional intercept abort priority. If provided, intercept is resolved using cooperative handling rules. Otherwise, intercept is resolved immediately.
```

});
```

#### httpRequest.continueRequestOverrides()

- returns: <[Object]> Optional request overwrites, which can be one of the following:

>   - `url` <[string]> If set changes the request URL. This is not a redirect. The request will be silently forwarded to the new URL. For example, the address bar will show the original URL.

Reviewer 2: ```suggestion
  - `url` <[string]> Changes the request URL if set. This is not a redirect — the request is silently forwarded to the new URL. For example, the address bar shows the original URL.
```

  - `method` <[string]> If set changes the request method (e.g. `GET` or `POST`).
  - `postData` <[string]> If set changes the post data of request.

>   - `headers` <[Object]> If set changes the request HTTP headers. Header values will be converted to a string.

Reviewer 2: ```suggestion
  - `headers` <[Object]> If set changes the request HTTP headers. Header values are converted to a string.
```

Contains the request's resource type as it was perceived by the rendering engine.
ResourceType will be one of the following: `document`, `stylesheet`, `image`, `media`, `font`, `script`, `texttrack`, `xhr`, `fetch`, `eventsource`, `websocket`, `manifest`, `other`.

#### httpRequest.respond(response, [priority])

- `response` <[Object]> Response that will fulfill this request
  - `status` <[number]> Response status code, defaults to `200`.
  - `headers` <[Object]> Optional response headers. Header values will be converted to a string.
  - `contentType` <[string]> If set, equals to setting `Content-Type` response header
  - `body` <[string]|[Buffer]> Optional response body

> - `priority` <[number]> - Optional intercept abort priority. If provided, intercept will be resolved using coopeative handling rules. Otherwise, intercept will be resovled immediately.

Reviewer 2: ```suggestion
- `priority` <[number]> - Optional intercept abort priority. If provided, intercept is resolved using cooperative handling rules. Otherwise, intercept is resolved immediately.
```
